swift/swift.py

```diff
@@ -52,102 +61,131 @@ def __init__(self, setup_env: dict, parent=None):
-            args = info.get("args", {})
-            app = cls(name, self, **args)
-            # set a slot of broadcast signal to router
-            app.broadcastRequested.connect(self._route_to_bus)
-            # add the app to the list of subscribers on each bus
-            for bus_name in info["bus"]:
-                self._subscribers[bus_name].append(app)
-            # store the app
-            self._apps[name] = app
-
-    def _show_frame(self, setup_app: dict):
-        """Shows frames of each app.
+        # create a bus
+        if "timeout" in info:
+            bus = Bus(name, info["timeout"])
+        else:
+            bus = Bus(name)
+        # set a slot of received signal to router
+        bus.received.connect(self._routeToApp)
+        bus.start()
+        # store the bus
+        self._buses[name] = bus
+        self._subscribers.setdefault(name, set())
+
+    def createApp(self, name: str, info: dict):
+        """Creates an app and shows their frames using set-up environment.
         
         Args:
-            setup_app: Set-up environment about app.
+            name: A name of app.
+            info: A dictionary containing app info. Each element is like below:
+              path: A path desired to be added for importing app.
+              module: A name of app module.
+              class: A name of app class.
+              show: Whether its frames is shown at the beginning.
+                True if you want to show the frames, otherwise False.
+              pos: A string that indicates the position of the frames.
+                It should be one of "left", "right", "top", or "bottom"; and is case-sensitive.
+                Otherwise, it will be regarded as default (AllDockWidgetAreas).
+              bus: A list of buses which app subscribes to.
+              args: Additional arguments for custom.
+                If there is no additional arguments, set None.
         """
-        pos_to_area = {
-            "left": Qt.LeftDockWidgetArea,
-            "right": Qt.RightDockWidgetArea,
-            "top": Qt.TopDockWidgetArea,
-            "bottom": Qt.BottomDockWidgetArea
-        }
-        for name, info in setup_app.items():
-            # show frames if the 'show' option is true
-            if info["show"]:
-                for frame in self._apps[name].frames():
-                    dockWidget = QDockWidget(name, self.mainWindow)
-                    dockWidget.setWidget(frame)
-                    area = pos_to_area.get(info["pos"], Qt.AllDockWidgetAreas)
-                    self.mainWindow.addDockWidget(area, dockWidget)
-        self.mainWindow.show()
+        # import the app module
+        with _add_to_path(os.path.dirname(info.get("path", ""))):
+            module = importlib.import_module(info["module"])
+        # create an app
+        cls = getattr(module, info["class"])
+        if "args" in info:
+            app = cls(name, self, **info["args"])

```

Comment:
Maybe `self` should be given as `parent=self` since we don't know the order..?

Reply:
No! It is defined in `app.BaseApp`.

```python
__init__(self, name: str, parent=None)
```

But, if you think it is ambiguous, I will change it into `parent=self`.

Reply:
I meant that `cls` which inherits `BaseApp` may take other arguments before `parent`, e.g.,
```python
class Foo(BaseApp):
  def __init__(self, name, arg1, arg2, parent=None):
    super().__init__(name, parent)
    ...
```

Reply:
Ah, I thought we restricted the order of arguments.

I will change it into `parent=self`.

Reply:
Since `parent` is an optional argument (which has a default value), we cannot force the subclasses to keep the order of (`name`, `parent=None`, `arg1`, `arg2`, ...).